### PR TITLE
Replace stubbed PERCENTILE_N with naive implementation

### PR DESCRIPTION
**Pull request by rkennedy-mode**
Replaces the previously stubbed PERCENTILE_N methods with a naive implementation. The naive version implements value interpolation following NumPy and Microsoft Excel, backed by an STL vector. This PR also circles back to `SetFunction.java` to introduce some input type checking that had previously been missing (this was discovered in the unit tests added for this change).

A large portion of this commit is due to importing the [Cereal 1.3.0](https://uscilab.github.io/cereal/) library for serialization of C++ objects. Feel free to ignore the files under `third_party/cpp/cereal` unless you feel like reviewing that library. I could have moved those files to a separate PR or commit, but I wanted to ensure a rollback would be atomic and not leave any bits behind.

This implementation works so long as there are no more than ~1M values (rows) in any VoltDB partition. Approaching and beyond that the implementation is likely to run into VoltDB errors due to the 8MB VARBINARY limits. A subsequent PR will introduce more efficient (likely at the expense of accuracy) means of implementing percentiles for those cases.

Partial resolution for mode/issues#2732.

### CODE REVIEW

File: src/hsqldb19b3/org/hsqldb_voltpatches/SetFunction.java [hunk 1]
```diff
@@ -547,7 +547,24 @@ static Type getType(int setType, Type type) {
             case OpTypes.PERCENTILE_75 :
             case OpTypes.PERCENTILE_95 :
             case OpTypes.PERCENTILE_99 :
-                return Type.SQL_DOUBLE;
+                switch (dataType) {
+                case Types.TINYINT :
```
Comment (zacMode) [inline src/hsqldb19b3/org/hsqldb_voltpatches/SetFunction.java:551]: missing indentation of nested cases

**Review by zacMode — COMMENTED** [review]
My eyes are burning from looking at C++, but otherwise looks good ;) 

One small formatting note inline. 

Also:

> This implementation works so long as there are no more than ~1M values (rows) in any VoltDB partition

Do we have away to detect this and fail early?

**Review by besquared — APPROVED** [review]
LGTM

File: src/hsqldb19b3/org/hsqldb_voltpatches/SetFunction.java (hunk 1 above)
Comment (rkennedy-mode) [reply src/hsqldb19b3/org/hsqldb_voltpatches/SetFunction.java:551]: If you look further up the source file, this appears to be VoltDB's style (not to indent `case` statements within `switch` statements). I prefer the style you're describing, but I'm trying to keep with the same style that VoltDB is using here.

Comment (rkennedy-mode) [reply src/hsqldb19b3/org/hsqldb_voltpatches/SetFunction.java:551]: lol…actually, I take it back. They're using different styles at different points in the same file. 🤷‍♂️

Comment (zacMode) [reply src/hsqldb19b3/org/hsqldb_voltpatches/SetFunction.java:551]: 👍 makes sense - sorry missed that. github being helpful and collapsing "unrelated blocks"

Comment (zacMode) [reply src/hsqldb19b3/org/hsqldb_voltpatches/SetFunction.java:551]: the fact that they do it _anywhere_, to me, moves this suggestion in to the "why bother" category..

**Review by mstockm — APPROVED** [review]
LGTM

**Comment by rkennedy-mode** [comment]
> > This implementation works so long as there are no more than ~1M values (rows) in any VoltDB partition
> 
> Do we have away to detect this and fail early?

Sort of. There's nothing in the Aggregate API to signal errors. There are other places in the code that throw C++ `SQLException`, which I assume is how this error gets returned whenever we try processing >1M rows:

> VOLTDB ERROR: SQL ERROR The size 16000016 of the value exceeds the size of the VARBINARY(8388608) column 

But C++ exceptions are fiddly things and if someone doesn't catch them (they're not declared like in Java, so it's not a compilation error to just throw one even if you don't warn callers that you might) they can lead to crashing the whole process (💀 RIP VoltDB), so I've been reluctant to use them without there being an explicit mention of it being okay to use them in API contracts.

In this case, if I were to add checks for this error we'd just be saving a bit of processing and memory before VoltDB itself throws the exception I mentioned above. The subsequent change to move us past 1M rows/partition would also replace any early exit code I'd be adding and that code should be written and deployed before we can release this feature to most/any customers. So I'm not too worried about releasing things as they are now.

**Pull request merged**